exchange/exchange.go

```diff
@@ -394,8 +396,10 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
 		bidResponseExt.Warnings[openrtb_ext.BidderReservedGeneral] = append(bidResponseExt.Warnings[openrtb_ext.BidderReservedGeneral], generalWarning)
 	}
 
+	e.bidValidationEnforcement.SetBidValidationStatus(r.Account.Validations)
```

Comment:
Written as is, `SetBidValidationStatus(account Validations)` does not modify the method's object but the return value `finalEnforcement`. Therefore, we are not setting a different value to the `host` (aka `e.bidValidationEnforcement`) unless we either:
Option 1. Modify `SetBidValidationStatus` to be a "setter" as suggested in a previous review:
```
646 - func (host Validations) SetBidValidationStatus(account Validations) Validations {
    + func (host Validations) SetBidValidationStatus(account Validations) {
647 -     finalEnforcement := host
648       if len(account.BannerCreativeMaxSize) > 0 {
649 -         finalEnforcement.BannerCreativeMaxSize = account.BannerCreativeMaxSize
    +         host.BannerCreativeMaxSize = account.BannerCreativeMaxSize
650       }
651 -     return finalEnforcement
652   }
config/config.go
```
Option 2. Assign the return value to `e.bidValidationEnforcement`:
```
391       for _, warning := range r.Warnings {
392           generalWarning := openrtb_ext.ExtBidderMessage{
393               Code:    errortypes.ReadCode(warning),
394               Message: warning.Error(),
395           }
396           bidResponseExt.Warnings[openrtb_ext.BidderReservedGeneral] = append(bidResponseExt.Warnings[openrtb_ext.BidderReservedGeneral], generalWarning)
397       }
398
399       e.bidValidationEnforcement.SetBidValidationStatus(r.Account.Validations)
    +     e.bidValidationEnforcement = e.bidValidationEnforcement.SetBidValidationStatus(r.Account.Validations)
400
401       // Build the response
402       return e.buildBidResponse(ctx, liveAdapters, adapterBids, r.BidRequestWrapper.BidRequest, adapterExtra, auc, bidResponseExt, cacheInstructions.return
403   }
exchange/exchange.go
```
Whichever pattern do you preffer is ok with me. The reason this bug was not getting reflected in the `exchangetests/` JSON tests is because we are setting the bid response size restrictions in both sides. Here's were it gets set in PBS config:
```
2391     bidValidation := config.Validations{}
2392     if bidValidationEnforcement == config.ValidationEnforce || bidValidationEnforcement == config.ValidationWarn {
2393         bidValidation = config.Validations{
2394             BannerCreativeMaxSize: bidValidationEnforcement,
2395             MaxCreativeWidth:      100,
2396             MaxCreativeHeight:     100,
2397         }
2398     }
2399
2400     return &exchange{
2401         adapterMap:               bidderAdapters,
2402         me:                       metricsConf.NewMetricsEngine(&config.Configuration{}, openrtb_ext.CoreBidderNames(), nil),
2403 *-- 13 lines: cache:                    &wellBehavedCache{},------------------------------------------------------------------
2416         server:                   config.Server{ExternalUrl: "http://hosturl.com", GvlID: 1, DataCenter: "Datacenter"},
2417         bidValidationEnforcement: bidValidation,
2418     }
2419 }

exchange/exchange_test.go
```
And then again at the account level:
```
2187     // Account Bid Validation Enforcement
2188     accountBidValidation := config.Validations{}
2189     if spec.BidValidationEnforcement == config.ValidationEnforce {
2190         accountBidValidation.BannerCreativeMaxSize = config.ValidationEnforce
2191     } else if spec.BidValidationEnforcement == config.ValidationWarn {
2192         accountBidValidation.BannerCreativeMaxSize = config.ValidationWarn
2193     }
2194
2195     auctionRequest := AuctionRequest{
2196         BidRequestWrapper: &openrtb_ext.RequestWrapper{BidRequest: &spec.IncomingRequest.OrtbRequest},
2197         Account: config.Account{
2198             ID:            "testaccount",
2199             EventsEnabled: spec.EventsEnabled,
2200             DebugAllow:    true,
2201             Validations:   accountBidValidation,
2202         },
2203         UserSyncs:     mockIdFetcher(spec.IncomingRequest.Usersyncs),
2204         ImpExtInfoMap: impExtInfoMap,
2205     }
2206
2207 *--  7 lines: if spec.StartTime > 0 {---------------------------------------------------------------------
2214
2215     bid, err := ex.HoldAuction(ctx, auctionRequest, debugLog)
exchange/exchange_test.go
```
Therefore, we weren't able to notice how the account validation settings where not getting copied to the validation enforcement object in line 399 of `exchange.go`. In order to correct, if you agree, I was thinking we could probably change the `BidValidationEnforcement` field from being a string, to a full `config.Validations` object along with the test framework modifications necessary.
```
4594   type exchangeSpec struct {
4595       GDPREnabled                bool                   `json:"gdpr_enabled"`
4596       IncomingRequest            exchangeRequest        `json:"incomingRequest"`
4597       OutgoingRequests           map[string]*bidderSpec `json:"outgoingRequests"`
4598       Response                   exchangeResponse       `json:"response,omitempty"`
4599       EnforceCCPA                bool                   `json:"enforceCcpa"`
4600       EnforceLMT                 bool                   `json:"enforceLmt"`
4601       AssumeGDPRApplies          bool                   `json:"assume_gdpr_applies"`
4602       DebugLog                   *DebugLog              `json:"debuglog,omitempty"`
4603       EventsEnabled              bool                   `json:"events_enabled,omitempty"`
4604       StartTime                  int64                  `json:"start_time_ms,omitempty"`
4605       BidIDGenerator             *mockBidIDGenerator    `json:"bidIDGenerator,omitempty"`
4606       RequestType                *metrics.RequestType   `json:"requestType,omitempty"`
4607       PassthroughFlag            bool                   `json:"passthrough_flag,omitempty"`
4608       HostSChainFlag             bool                   `json:"host_schain_flag,omitempty"`
4609 -     BidValidationEnforcement   string                 `json:"bid_validation_flag,omitempty"`
     +     pbsConfigBidValidation     config.Validations     `json:"config_bid_validations"`
     +     accountConfigBidValidation config.Validations     `json:"account_bid_validations"`
4610   }
exchange/exchange_test.go
```
Under this logic, we could probably have JSON test where the account bid validation config overrides that of PBS. An extra, is the ability to set the max values in the JSON test itself. For instance:
```
1   {
2 -     "bid_validation_flag": "enforce",
  +     "description": "Bid request validation enforcement at both system and account level. Expect PBS values being overriden by account-level values"
  +     "config_bid_validations": {
  +         "banner_creative_max_size": "warn",
  +         "secure_markup": "skip",
  +         "max_creative_width": 600,
  +         "max_creative_height": 480
  +     },
  +     "account_bid_validations": "enforce",
  +         "banner_creative_max_size": "enforce",
  +     },
3       "incomingRequest": {
4           "ortbRequest": {
5               "id": "some-request-id",
6               "site": {
exchange/exchangetest/bid_response_validation_enforce_one_bid_rejected.json
```

Reply:
This is a great catch Gus! In my latest commit I'll have an update for these end-to-end tests to better check for this along with updating the SetBidValidationStatus to properly set the host config.

Reply:
I also updated the end-to-end tests in general to support checking that the error messages are being written properly.